app/assets/javascripts/student_profile/TransitionNotes.js

```diff
@@ -133,7 +98,7 @@ class TransitionNotes extends React.Component {
   autosaveStatusText(status) {
     if (status === 'saved') return 'Saved.';
```

Comment:
@alexsoble Here's two thoughts on simplifying this.  I think what you have here is a great improvement, so I think it completely makes sense to 🚢 this, and it's up to you if you think these other ideas are worth doing now or after or not at all.

First, this has two separate "threads" running that can modify the state on the server, one for the normal note and the other for the restricted note.  And the throttling is happening on a fn that's processing the state changes for the different text boxes - a few layers away from the actual server call.  This invites race conditions, since there are two separate ways that can make server calls, and they're throttled independently, rather than adding the throttling around the server call directly.  If the user modifies one text box it fires off a request, and then if they modify the second text box it fires off another request.  It'd be simpler to understand the communication with the server if you could throttle a single function that is communicating with the server.  Here, that might mean adding a `doSave` method, updating the other methods so that they call that instead of calling `this.props.onSave(params)` directly, and then only throttle `doSave`.

Second, as you described in your awesome PR description, the problem is related to when there are multiple requests.  Tracking only the state of the latest request isn't sufficient for understanding what's happening there, just like you describe.  This works in other places in the UI like when you're taking notes, because the UI essentially blocks everything to make sure there can only be one server request in-flight at once.

So instead of storing the state of the request to the server, another approach is to store "what does the client think the state of the server is?"  And then if that's different than the client state, you know there's unsaved changes.  Here that could mean instead of tracking the last change to any request state, track what the `lastServerState` is.  On load it's the same as whatever data is read from the server.  Then that's the value to check to see if the client thinks everything is saved.  If there are multiple requests to the server, and they go out and come back in out of order, this value gets updated whenever a response from the server comes back.  This still isn't perfect (since requests could finish in order on the server, but come back to the client out of order), but is probably good enough for most normal purposes.

Reply:
Woah, this a fantastic way to change how I think about tracking state! This helps me see the advantages of some of the changes you suggested earlier, like consolidating down to one `doSave` call. I won't implement this today since it involves big API changes, and I want to get an improved UX version of this feature to Les by today. But I definitely want to implement this pattern either here or in the next component I write that involves sharing multiple streams of information between client/server!